Directory.EnumerateDirectories doesn't return directory list on rhel7.2

_From @jainaashish on April 7, 2018 0:18_

I've created a simple application to demonstrate the issue where `Directory.EnumerateDirectories` api doesn't return the list of child directories on rhel 7.2, although it works fine on Windows or Ubuntu. 

Sample Project - https://github.com/jainaashish/CoreClrIssue/tree/master/EnumerateDirectoriesIssue/EnumerateDirectoriesIssue


_Copied from original issue: dotnet/coreclr#17465_

@JeremyKuhne I just checked on rhel7 and its Directory.Enumerate Directories are returning the list of directories for corefx repo.
Test output
```
  Discovering: System.IO.FileSystem.Tests
  Discovered:  System.IO.FileSystem.Tests
  Starting:    System.IO.FileSystem.Tests
     System.IO.Tests.EnumerableTests.EnumerateDirectories_Rhel [FAIL]
        Assert.Equal() Failure
        Expected: String[] []
        Actual:   FileSystemEnumerable<String> ["/tmp/EnumerableTests_0olhl43x.apn/EnumerateDirecto"..., "/tmp/EnumerableTests_0olhl43x.apn/EnumerateDirecto"...]
        Stack Trace:
           /git/corefx/src/System.IO.FileSystem/tests/Directory/EnumerableTests.cs(54,0): at System.IO.Tests.EnumerableTests.EnumerateDirectories_Rhel()
  Finished:    System.IO.FileSystem.Tests

  === TEST EXECUTION SUMMARY ===
/git/corefx/Tools/tests.targets(492,5): warning :    System.IO.FileSystem.Tests  Total: 3993, Errors: 0, Failed: 1, Skipped: 0, Time: 25.061s [/git/corefx/src/System.IO.FileSystem/tests/System.IO.FileSystem.Tests.csproj]
  /git/corefx/src/System.IO.FileSystem/tests
  ----- end 15:31:59 ----- exit code 1 ----------------------------------------------------------
  Looking around for any Linux dump...
  ... found no dump in /git/corefx/bin/tests/System.IO.FileSystem.Tests/netcoreapp-Linux-Debug-x64
/git/corefx/Tools/tests.targets(492,5): warning MSB3073: The command "/git/corefx/bin/tests/System.IO.FileSystem.Tests/netcoreapp-Linux-Debug-x64/RunTests.sh /git/corefx/bin/testhost/netcoreapp-Linux-Debug-x64/" exited with code 1. [/git/corefx/src/System.IO.FileSystem/tests/System.IO.FileSystem.Tests.csproj]
/git/corefx/Tools/tests.targets(500,5): error : One or more tests failed while running tests from 'System.IO.FileSystem.Tests' please check /git/corefx/bin/tests/System.IO.FileSystem.Tests/netcoreapp-Linux-Debug-x64/testResults.xml for details! [/git/corefx/src/System.IO.FileSystem/tests/System.IO.FileSystem.Tests.csproj]
```

Here is the version for my rhel machine 
```
[root@75ea72ed8713 corefx]# cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.2 (Maipo)
[root@75ea72ed8713 corefx]#
```

--- a/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs
+++ b/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs
@@ -44,10 +44,14 @@ namespace System.IO.Enumeration
                 // We know it's a directory.
                 isDirectory = true;
             }
-            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK)
-                && Interop.Sys.Stat(entry.FullPath, out Interop.Sys.FileStatus targetStatus) >= 0)
+            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK
+                || directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN)
+                && (Interop.Sys.Stat(entry.FullPath, out Interop.Sys.FileStatus targetStatus) >= 0
+                    || Interop.Sys.LStat(entry.FullPath, out targetStatus) >= 0))
             {
-                // It's a symlink: stat to it to see if we can resolve it to a directory.
+                // Symlink or unknown: Stat to it to see if we can resolve it to a directory. If Stat fails,
+                // it could be because the symlink is broken, we don't have permissions, etc., in which
+                // case fall back to using LStat to evaluate based on the symlink itself.
                 isDirectory = (targetStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR;
             }
 

